### DISCUSSION TITLE
Batch operations (nodeJS)

### DISCUSSION THREAD

**Discussion by suelitonx** [pocketbase/pocketbase#6040] [Q&A] [answered]
Could someone help me? Batch requests are taking too long (10,000 records are taking 5 minutes). I split the batch operations into chunks of 1,000 records to avoid timeout issues. I also increased the batch request limit and the timeout settings, but it didn’t help.

```
import fs from "fs/promises";
import PocketBase from "pocketbase";

async function processJsonInBatches(filePath, collectionName, batchSize = 1000) {
    
    const pb = new PocketBase("http://localhost:7777");
    await pb.collection("_superusers").authWithPassword("suelitonx@gmail.com", "pocketbase123456");

    const fileContent = await fs.readFile(filePath, "utf8");
    const data = JSON.parse(fileContent);

    const totalBatchCount = Math.ceil(data.length / batchSize);
    
    let errors = 0;

    for (let i = 0; i < data.length; i += batchSize) {

      const batchData = data.slice(i, i + batchSize);
      
      const batch = pb.createBatch();

      for (const record of batchData) {
        batch.collection(collectionName).create(record);
      }

      try {
        await batch.send();
      } catch (_) {
        errors++;
      }

      console.log(`Batch ${Math.floor(i / batchSize) + 1} / ${totalBatchCount} processed successfully. Records: ${batchData.length - errors}/ ${batchData.length}.`);
    }

    
}

(async () => {
    try {
        await processJsonInBatches("C:/file.json", "collection_name");
    } catch (error) {
        console.error("Processing failed:", error);
    }
})();
```

**Comment by ganigeorgiev** [comment, marked as answer]
It would help to provide sample data to understand what you are inserting (_e.g. what is the size of a single record payload, what are the fields you are targeting, etc._) and to elaborate whether you have any custom hooks that could affect the create operation (_e.g. sending an email or some other slow operation_).

It is not even clear whether the slow time you are experiencing is actually caused by the request processing or something else related to your JS code.

In any case, as mentioned in your other post, the batch web API is not intended to be used this way and it is expected to be slower when compared to a direct db insert because it not only needs to send the data over the network and then return it back, it also performs a lot of internal copying, serialization/deserialization, validations, etc.

If all you are trying to do is to import a json file into your database, then [registering a custom CLI](https://pocketbase.io/docs/js-console-commands/) would be a more optimal solution. Here is a minimal `pb_hooks/main.pb.js` example:

```js
/// <reference path="../pb_data/types.d.ts" />

$app.rootCmd.addCommand(new Command({
    use: "import",
    run: (cmd, args) => {
        const items = require(args[0])

        const collection = $app.findCollectionByNameOrId(args[1])

        $app.runInTransaction((txApp) => {
            let record;
            for (let item of items) {
                record = new Record(collection)
                record.load(item)
                // call txApp.saveNoValidate to increase further the performance
                // but be aware that the fields validators will be skipped and could result in data integrity issues
                // if you haven't checked and verified the imported data
                txApp.save(record)
            }
        })
    },
}))
```
To run the custom command: `./pocketbase import /path/to/your/file.json yourCollection`.
